### PR TITLE
204 NLP Document Inference Entity Extraction Demo

### PR DESCRIPTION
**Pull request by rageshhajela16**
This notebook demo shows NLP-based named entity extraction.
Inference is performed on multiple examples to extract different entities on simple text.

### CODE REVIEW

**Comment by rageshhajela16** [comment]
Thanks @adrianboguszewski for the initial review on draft pull request https://github.com/openvinotoolkit/openvino_notebooks/pull/418 . Here are the proposed changes in this new pull request, please review and let me know how to improve further or any other comments.

1. Changed the name to 204
2.  Added a link to the notebook to the main README file
3. I have moved pytesseract calls in if-blocks or handled exception gracefully. Let me know if it serves.

Didn't observe any merge conflicts or CI failures and conflicts, will wait for your feedback if needs some fix.
I observed few messages titled "Some checks haven’t completed yet", not sure how to proceed on that, please suggest.

File: notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import cv2\n",
+    "import time\n",
+    "import json\n",
+    "import operator\n",
+    "from urllib import parse\n",
```
Comment (adrianboguszewski) [inline notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb:41]: This import is not used

File: notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+    "            \"pytesseract.pytesseract.tesseract_cmd to the directory that contains tesseract.exe \"\n",
+    "        )\n",
+    "    else:\n",
+    "        pytesseract.pytesseract.tesseract_cmd=str(ocr_path[-1])\n",
+    "        print(f\"Successfully set pytesseract.pytesseract.tesseract_cmd to path {ocr_path[-1]}\\n\")\n",
+    "        \n",
+    "    # Test converting sample JPG image into string using pytesseract OCR\n",
+    "    print('Extracted OCR text:')\n",
+    "    img = cv2.imread('data/sample.jpg')\n",
+    "    print(pytesseract.image_to_string(img))"
```
Comment (adrianboguszewski) [inline notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb:91]: These lines should be under else to deal with the situation you don't have tesseract installed on Windows

File: notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+   "source": [
+    "healthcare_template = [\"name\", \"age\", \"medical condition\", \"medication\", \"dosage\", \"dosage unit\"]\n",
+    "\n",
+    "def run_analyze_entities(source, source_type = \"text\"):\n",
+    "    print(f\"Context: {source}\\n\", flush=True)\n",
+    "    \n",
+    "    try:\n",
+    "        context = load_context(source, source_type)\n",
+    "    except:\n",
+    "        print(\"Error: OCR issue in reading file and loading context\")\n",
```
Comment (adrianboguszewski) [inline notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb:491]: This clause is too broad. It's better to catch TesseractNotFoundError. Also the message should be more meaningful like "To extract entities from image you need to install Tesseract-OCR engine". Maybe also some link how to install it?

**Review by adrianboguszewski — CHANGES_REQUESTED** [review]
Thanks @rageshhajela16. It's getting better and better. I think we need only a few changes to this notebook.

File: notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb (hunk 3 above)
Comment (rageshhajela16) [reply notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb:491]: @adrianboguszewski Thanks, makes sense. Is it fine if I modify the message as follows?

        print("OCR Error: To extract entities from image you need to install Tesseract-OCR engine.
              "Follow the installation steps mentioned here "
              "https://stackoverflow.com/questions/50655738/how-do-i-resolve-a-tesseractnotfounderror")

Comment (rageshhajela16) [reply notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb:491]: As TesseractNotFoundError belong to this class pytesseract.pytesseract.TesseractNotFoundError, so I created a custom exception of the same name to raise at OCR call. And then caught this exception TesseractNotFoundError during entity analysis. Please review if this looks good, thanks!

**Comment by adrianboguszewski** [comment]
I'm ok with the changes but you need also to pass all CI jobs. Please see the reasons and fix them.

**Comment by rageshhajela16** [comment]
> I'm ok with the changes but you need also to pass all CI jobs. Please see the reasons and fix them.

Sure, thanks for the review @adrianboguszewski . Next I am on it to fix the CI issues and will update soon.

**Comment by rageshhajela16** [comment]
sample

**Review by adrianboguszewski — CHANGES_REQUESTED** [review]
Thanks @rageshhajela16 for restoring dynamic shapes. We found that they don't work correctly if the input is too long. It shouldn't be a problem with this notebook (as the input is rather short) but it is desirable to have security. Hence
- please process context only if is shorter than 380 words, if it's longer print the message "The context is too long for this particular model. Try with context shorter than 380 words."
- please fix all CI issues. All CI must be green before we can merge
- please change the template to: name, age, address, phone number and so on
- description must fit the code

**Comment by ryanloney** [comment]
Can you use a PDF document from intel.com? Something like this that can be downloaded inside the notebook:

https://download.intel.com/newsroom/2021/manufacturing/12th-Gen-Intel-Core-SoC-Overview.pdf

**Review by Debskij — CHANGES_REQUESTED** [review]

**Comment by rageshhajela16** [comment]
> Can you use a PDF document from intel.com? Something like this that can be downloaded inside the notebook:
> 
> https://download.intel.com/newsroom/2021/manufacturing/12th-Gen-Intel-Core-SoC-Overview.pdf

Thanks @ryanloney, I tried using the referred pdf, but since it's having larger context, dynamic shapes which has been restored on this notebook might be at risk. Also since this PDF has embedded images, straight forward PDF to text extraction using PdfFileReader or pdfplumber was challenging.

As a workaround, I have tried to use more generic sentences in form of text, images and PDFs and uploaded the PDF file with approximately 19 KBs size. Please review proposed changes and let me know if it would work out. Thanks.

File: notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+    "for input_layer in model.inputs:\n",
+    "    input_shape = input_layer.partial_shape\n",
+    "    input_shape[1] = -1\n",
+    "    model.reshape({input_layer: input_shape})\n",
+    "# compile the model for the CPU\n",
+    "compiled_model = ie_core.compile_model(model=model, device_name=\"CPU\")\n",
+    "\n",
+    "# get input and output names of nodes\n",
+    "input_keys = list(compiled_model.inputs)\n",
+    "output_keys = list(compiled_model.outputs)"
```
[the remark this answers is not in the document]
Comment (rageshhajela16) [reply notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb]: Thanks for the suggestion. Usage of redundant output_keys is omitted.

File: notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "[i.any_name for i in input_keys], [o.any_name for o in output_keys]"
```
[the remark this answers is not in the document]
Comment (rageshhajela16) [reply notebooks/204-document-entity-extraction/204-document-entity-extraction.ipynb]: Thanks @Debskij for the suggestion. Print of redundant output_keys is omitted based on previous comment.

**Review by adrianboguszewski — APPROVED** [review]

File: .docker/Pipfile [hunk 6]
```diff
@@ -56,4 +56,4 @@ yaspin = "*"
 [dev-packages]
 
 [requires]
-python_version = "3.8"
+python_version = "3.8"
```
Comment (Debskij) [inline .docker/Pipfile:59]: Please revert changes to current main status for Pipfile and Pipfile.lock

File: .docker/Pipfile.lock [hunk 7]
```diff
@@ -4042,4 +4042,4 @@
         }
     },
     "develop": {}
-}
+}
```
Comment (Debskij) [inline .docker/Pipfile.lock:4045]: Please revert changes to current main status for Pipfile and Pipfile.lock

**Review by Debskij — CHANGES_REQUESTED** [review]

**Review by raymondlo84 — APPROVED** [review]
Let's do it.

File: .docker/Pipfile (hunk 6 above)
Comment (raymondlo84) [reply .docker/Pipfile:59]: Can we just override these?

Comment (rageshhajela16) [reply .docker/Pipfile]: Thanks for the tips @Debskij @raymondlo84 , I have overwritten these files and removed from this pull request.

File: .docker/Pipfile.lock (hunk 7 above)
Comment (rageshhajela16) [reply .docker/Pipfile.lock]: I have overwritten these files and removed from this pull request.

**Review by raymondlo84 — APPROVED** [review]

**Review by raymondlo84 — APPROVED** [review]

**Review by Debskij — APPROVED** [review]

**Comment by raymondlo84** [comment]
Magic... finally done! Thanks everybody.

**Pull request merged**